codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamWriteDistributor.java

```diff
@@ -0,0 +1,425 @@
+                childState.allocate(bytesForChild);
+                bytesAllocated += bytesForChild;
+                connectionWindow -= bytesForChild;
+            }
+            int childBytesAllocated = allocateBytesForTree(child, connectionWindow);
+            bytesAllocated += childBytesAllocated;
+            connectionWindow -= childBytesAllocated;
+            return true;
+        }
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class PriorityState {
+        final Http2Stream stream;
+        boolean hasFrame;
+        int streamableBytes;
+        int allocated;
+        int unallocatedStreamableBytesForTree;
+
+        PriorityState(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        /**
+         * Recursively increments the {@link #unallocatedStreamableBytesForTree()} for this branch in
+         * the priority tree starting at the current node.
+         */
+        void unallocatedStreamableBytesForTreeChanged(int delta) {
+            unallocatedStreamableBytesForTree += delta;
+            if (!stream.isRoot()) {
+                state(stream.parent()).unallocatedStreamableBytesForTreeChanged(delta);
+            }
+        }
+
+        void allocate(int bytes) {
+            allocated += bytes;
+
+            // Also artificially reduce the streamable bytes for this tree to give the appearance
+            // that the data has been written. This will be restored before the allocated bytes are
+            // actually written.
+            unallocatedStreamableBytesForTreeChanged(-bytes);
+        }
+
+        /**
+         * Reset the number of bytes that have been allocated to this stream by the priority
+         * algorithm.
+         */
+        void resetAllocated() {
+            allocate(-allocated);
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean hasFrame) {
+            this.hasFrame = hasFrame;
+
+            int delta = newStreamableBytes - streamableBytes;
+            streamableBytes = newStreamableBytes;
+
+            // Update this branch of the priority tree if the streamable bytes have changed for this node.
+            unallocatedStreamableBytesForTreeChanged(delta);
+        }
+
+        void close() {
+            // Unallocate all bytes.
+            resetAllocated();
+
+            // Clear the streamable bytes.
+            updateStreamableBytes(0, false);
+        }
+
+        boolean hasFrame() {
+            return hasFrame;
+        }
+
+        int unallocatedStreamableBytes() {
+            return streamableBytes - allocated;
+        }
+
+        int unallocatedStreamableBytesForTree() {
+            return unallocatedStreamableBytesForTree;
+        }
+    }
+
+    /**
+     * A connection stream visitor that delegates to the user provided visitor.
+     */
+    private class WriteVisitor implements Http2StreamVisitor {
```

Comment:
@nmittler - I think we have an opportunity to simplify the state management and iteration mechanism currently employed in `PriorityStreamWriteDistributor`. Instead of explaining it I have a diff that you can apply and take a look...let me know what you think.

```diff
diff --git a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java b/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java
index 6a6ea7e..c40f272 100644
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java
@@ -62,6 +62,7 @@ public class DefaultHttp2RemoteFlowController implements Http2RemoteFlowControll
                                             StreamWriteDistributor streamWriteDistributor) {
         this.connection = checkNotNull(connection, "connection");
         this.streamWriteDistributor = checkNotNull(streamWriteDistributor, "streamWriteDistributor");
+        streamWriteDistributor.writer(WRITER);
 
         // Add a flow state for the connection.
         stateKey = connection.newKey();
@@ -275,7 +276,7 @@ public class DefaultHttp2RemoteFlowController implements Http2RemoteFlowControll
         int bytesToWrite = writableBytes(connectionState().windowSize());
 
         // Distribute the connection window across the streams and write the data.
-        streamWriteDistributor.distributeWrites(bytesToWrite, WRITER);
+        streamWriteDistributor.distributeWrites(bytesToWrite);
     }
 
     /**
diff --git a/codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamWriteDistributor.java b/codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamWriteDistributor.java
index 410ea7a..44cd3e3 100644
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamWriteDistributor.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamWriteDistributor.java
@@ -21,14 +21,18 @@ import static java.lang.Math.min;
 
 import java.util.Arrays;
 
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
 /**
  * A {@link StreamWriteDistributor} that implements the HTTP/2 priority tree algorithm for allocating
  * bytes for all streams in the connection.
  */
 public final class PriorityStreamWriteDistributor implements StreamWriteDistributor {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(PriorityStreamWriteDistributor.class);
     private final Http2Connection connection;
     private final Http2Connection.PropertyKey stateKey;
-    private final WriteVisitor writeVisitor = new WriteVisitor();
+    private Http2StreamVisitor writeVisitor;
 
     public PriorityStreamWriteDistributor(Http2Connection connection) {
         this.connection = checkNotNull(connection, "connection");
@@ -74,6 +78,28 @@ public final class PriorityStreamWriteDistributor implements StreamWriteDistribu
         });
     }
 
+    @Override public void writer(final Writer writer) {
+        checkNotNull(writer, "writer");
+        writeVisitor = new Http2StreamVisitor() {
+            @Override
+            public boolean visit(Http2Stream stream) throws Http2Exception {
+                PriorityState state = state(stream);
+                int allocated = state.allocated;
+
+                // Unallocate all bytes for this stream.
+                state.resetAllocated();
+
+                // Write the allocated bytes.
+                try {
+                    writer.write(stream, allocated);
+                } catch (RuntimeException e) {
+                    logger.error("unexpected error caught from", e);
+                }
+                return true;
+            }
+        };
+    }
+ 
     @Override
     public void updateStreamableBytes(StreamState streamState) {
         state(streamState.stream()).updateStreamableBytes(streamState.streamableBytes(),
@@ -81,14 +107,13 @@ public final class PriorityStreamWriteDistributor implements StreamWriteDistribu
     }
 
     @Override
-    public boolean distributeWrites(int maxBytes, Writer writer) {
-        checkNotNull(writer, "writer");
+    public boolean distributeWrites(int maxBytes) throws Http2Exception {
         if (maxBytes > 0) {
             allocateBytesForTree(connection.connectionStream(), maxBytes);
         }
 
         // Need to write even if maxBytes == 0 in order to handle the case of empty frames.
-        writeVisitor.writeAllocatedBytes(writer);
+        connection.forEachActiveStream(writeVisitor);
 
         return state(connection.connectionStream()).unallocatedStreamableBytesForTree() > 0;
     }
@@ -368,58 +393,4 @@ public final class PriorityStreamWriteDistributor implements StreamWriteDistribu
             return unallocatedStreamableBytesForTree;
         }
     }
-
-    /**
-     * A connection stream visitor that delegates to the user provided visitor.
-     */
-    private class WriteVisitor implements Http2StreamVisitor {
-        Writer writer;
-        RuntimeException error;
-        boolean writing;
-
-        void writeAllocatedBytes(Writer writer) {
-            assert !writing;
-            writing = true;
-            try {
-                error = null;
-                this.writer = writer;
-                try {
-                    connection.forEachActiveStream(this);
-                } catch (Http2Exception e) {
-                    // Should never happen since the visitor doesn't throw.
-                    throw new IllegalStateException(e);
-                }
-
-                // If an error was caught when calling back the visitor, throw it now.
-                if (error != null) {
-                    throw error;
-                }
-            } finally {
-                writing = false;
-            }
-        }
-
-        @Override
-        public boolean visit(Http2Stream stream) {
-            PriorityState state = state(stream);
-            try {
-                int allocated = state.allocated;
-
-                // Unallocate all bytes for this stream.
-                state.resetAllocated();
-
-                // Write the allocated bytes.
-                if (error == null) {
-                    writer.write(stream, allocated);
-                }
-            } catch (RuntimeException e) {
-                // Stop calling the visitor, but continue in the loop to reset the allocated for
-                // all remaining states.
-                error = e;
-            }
-
-            // We have to iterate across all streams to ensure that we reset the allocated bytes.
-            return true;
-        }
-    }
 }
diff --git a/codec-http2/src/main/java/io/netty/handler/codec/http2/StreamWriteDistributor.java b/codec-http2/src/main/java/io/netty/handler/codec/http2/StreamWriteDistributor.java
index 7064e3a..b67c0d2 100644
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/StreamWriteDistributor.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/StreamWriteDistributor.java
@@ -66,6 +66,11 @@ public interface StreamWriteDistributor {
     void updateStreamableBytes(StreamState state);
 
     /**
+     * Set the writer to use in {@link #distributeWrites(int)}.
+     */
+    void writer(Writer writer);
+
+    /**
      * Distributes up to {@code maxBytes} to those streams containing streamable bytes and
      * iterates across those streams to write the appropriate bytes. Criteria for
      * traversing streams is undefined and it is up to the implementation to determine when to stop
@@ -79,5 +84,5 @@ public interface StreamWriteDistributor {
      * @return {@code true} if there are still streamable bytes that have not yet been written,
      * otherwise {@code false}.
      */
-    boolean distributeWrites(int maxBytes, Writer writer);
+    boolean distributeWrites(int maxBytes) throws Http2Exception;
 }
```